### PR TITLE
Further optimise FastRegexMatcher

### PR DESCRIPTION
**Pull request by pracucci**
In Grafana Mimir we're using a modified version of `FastRegexMatcher` which evolved over the time to cover more and more cases we've seen in production. Whenever we see query performance degradation introduced by regexp matcher applied to a large number of values, we get back to `FastRegexMatcher` to see if there's any further optimization can be done.

Over the time we've accumulated several optimizations and the code complexity increased significantly. Few days ago I've [asked on Slack](https://cloud-native.slack.com/archives/C01AUBA4PFE/p1705995672177009) if there was any interest for Prometheus community to get these optimizations too, and Ben and Bartek have been open to look at them. At that time, Bartek mentioned he would have preferred a big bag PR instead of several incremental ones, but we can eventually revisit this decision (I understand this PR is quite tricky to review, being the result of a couple of years of iterations we've done in Mimir).

### How many regexp matchers are optimized in the wild?

In Grafana Mimir we track a metric to keep monitored the % of optimized regexp matchers. To give you a data point, over the last 7 days the 97.2% of matchers were running through the optimization (so for these matchers we didn't touch the regex engine at all).

Ref query:

```
sum(rate(cortex_query_frontend_regexp_matcher_optimized_count[7d]))
/
sum(rate(cortex_query_frontend_regexp_matcher_count[7d]))
```

### Benchmark

The new `NewFastRegexMatcher()` is slower because more cases are evaluated, so some previously optimized cases show a regression. However, such regressions are very small in absolute terms (e.g. few tens of nanoseconds) while the new optimizations for typical slow queries are pretty large (especially in absolute terms).

```
goos: darwin
goarch: amd64
pkg: github.com/prometheus/prometheus/model/labels
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
                                                                                                             │   before.txt   │              after.txt              │
                                                                                                             │     sec/op     │    sec/op     vs base               │
FastRegexMatcher/#00-12                                                                                          42.17n ± ∞ ¹   66.40n ± ∞ ¹  +57.46% (p=0.008 n=5)
FastRegexMatcher/foo-12                                                                                          42.07n ± ∞ ¹   80.09n ± ∞ ¹  +90.37% (p=0.008 n=5)
FastRegexMatcher/^foo-12                                                                                        120.00n ± ∞ ¹   65.60n ± ∞ ¹  -45.33% (p=0.008 n=5)
FastRegexMatcher/(foo|bar)-12                                                                                  1008.00n ± ∞ ¹   73.88n ± ∞ ¹  -92.67% (p=0.008 n=5)
FastRegexMatcher/foo.*-12                                                                                        763.1n ± ∞ ¹   157.9n ± ∞ ¹  -79.31% (p=0.008 n=5)
FastRegexMatcher/.*foo-12                                                                                        694.5n ± ∞ ¹   170.0n ± ∞ ¹  -75.52% (p=0.008 n=5)
FastRegexMatcher/^.*foo$-12                                                                                      712.5n ± ∞ ¹   168.6n ± ∞ ¹  -76.34% (p=0.008 n=5)
FastRegexMatcher/^.+foo$-12                                                                                      701.5n ± ∞ ¹   166.5n ± ∞ ¹  -76.27% (p=0.008 n=5)
FastRegexMatcher/.*-12                                                                                          6324.0n ± ∞ ¹   134.3n ± ∞ ¹  -97.88% (p=0.008 n=5)
FastRegexMatcher/.+-12                                                                                          6115.0n ± ∞ ¹   133.0n ± ∞ ¹  -97.83% (p=0.008 n=5)
FastRegexMatcher/foo.+-12                                                                                        749.1n ± ∞ ¹   159.8n ± ∞ ¹  -78.67% (p=0.008 n=5)
FastRegexMatcher/.+foo-12                                                                                        689.9n ± ∞ ¹   168.4n ± ∞ ¹  -75.59% (p=0.008 n=5)
FastRegexMatcher/foo_.+-12                                                                                       114.8n ± ∞ ¹   145.2n ± ∞ ¹  +26.48% (p=0.008 n=5)
FastRegexMatcher/foo_.*-12                                                                                       114.5n ± ∞ ¹   145.4n ± ∞ ¹  +26.99% (p=0.008 n=5)
FastRegexMatcher/.*foo.*-12                                                                                     2297.0n ± ∞ ¹   301.8n ± ∞ ¹  -86.86% (p=0.008 n=5)
FastRegexMatcher/.+foo.+-12                                                                                     2214.0n ± ∞ ¹   307.7n ± ∞ ¹  -86.10% (p=0.008 n=5)
FastRegexMatcher/(?s:.*)-12                                                                                    5728.00n ± ∞ ¹   61.75n ± ∞ ¹  -98.92% (p=0.008 n=5)
FastRegexMatcher/(?s:.+)-12                                                                                    5600.00n ± ∞ ¹   79.03n ± ∞ ¹  -98.59% (p=0.008 n=5)
FastRegexMatcher/(?s:^.*foo$)-12                                                                                 739.8n ± ∞ ¹   161.9n ± ∞ ¹  -78.12% (p=0.008 n=5)
FastRegexMatcher/(?i:foo)-12                                                                                     877.0n ± ∞ ¹   114.2n ± ∞ ¹  -86.98% (p=0.008 n=5)
FastRegexMatcher/(?i:(foo|bar))-12                                                                               993.9n ± ∞ ¹   248.9n ± ∞ ¹  -74.96% (p=0.008 n=5)
FastRegexMatcher/(?i:(foo1|foo2|bar))-12                                                                        1006.0n ± ∞ ¹   436.8n ± ∞ ¹  -56.58% (p=0.008 n=5)
FastRegexMatcher/^(?i:foo|oo)|(bar)$-12                                                                          1.010µ ± ∞ ¹   1.084µ ± ∞ ¹   +7.33% (p=0.008 n=5)
FastRegexMatcher/(?i:(foo1|foo2|aaa|bbb|ccc|ddd|e-12                                                             9.922µ ± ∞ ¹   2.022µ ± ∞ ¹  -79.62% (p=0.008 n=5)
FastRegexMatcher/((.*)(bar|b|buzz)(.+)|foo)$-12                                                                11307.0n ± ∞ ¹   709.6n ± ∞ ¹  -93.72% (p=0.008 n=5)
FastRegexMatcher/^$-12                                                                                          782.30n ± ∞ ¹   65.81n ± ∞ ¹  -91.59% (p=0.008 n=5)
FastRegexMatcher/(prometheus|api_prom)_api_v1_.+-12                                                              212.9n ± ∞ ¹   228.5n ± ∞ ¹   +7.33% (p=0.008 n=5)
FastRegexMatcher/10\.0\.(1|2)\.+-12                                                                              114.9n ± ∞ ¹   142.9n ± ∞ ¹  +24.37% (p=0.008 n=5)
FastRegexMatcher/10\.0\.(1|2).+-12                                                                               114.1n ± ∞ ¹   143.0n ± ∞ ¹  +25.33% (p=0.008 n=5)
FastRegexMatcher/((fo(bar))|.+foo)-12                                                                           8677.0n ± ∞ ¹   297.5n ± ∞ ¹  -96.57% (p=0.008 n=5)
FastRegexMatcher/zQPbMkNO|NNSPdvMi|iWuuSoAl|qbvKM-12                                                           32289.0n ± ∞ ¹   288.0n ± ∞ ¹  -99.11% (p=0.008 n=5)
FastRegexMatcher/jyyfj00j0061|jyyfj00j0062|jyyfj9-12                                                            4406.0n ± ∞ ¹   304.6n ± ∞ ¹  -93.09% (p=0.008 n=5)
FastRegexMatcher/(?i:(zQPbMkNO|NNSPdvMi|iWuuSoAl|-12                                                            49.893µ ± ∞ ¹   2.031µ ± ∞ ¹  -95.93% (p=0.008 n=5)
FastRegexMatcher/(?i:(zQPbMkNO.*|NNSPdvMi.*|iWuuS-12                                                            55.363µ ± ∞ ¹   9.868µ ± ∞ ¹  -82.18% (p=0.008 n=5)
FastRegexMatcher/(?i:(.*zQPbMkNO|.*NNSPdvMi|.*iWu-12                                                           1655.29µ ± ∞ ¹   11.87µ ± ∞ ¹  -99.28% (p=0.008 n=5)
FastRegexMatcher/fo.?-12                                                                                         172.6n ± ∞ ¹   149.9n ± ∞ ¹  -13.15% (p=0.008 n=5)
FastRegexMatcher/foo.?-12                                                                                        172.2n ± ∞ ¹   149.3n ± ∞ ¹  -13.30% (p=0.008 n=5)
FastRegexMatcher/f.?o-12                                                                                         120.4n ± ∞ ¹   146.3n ± ∞ ¹  +21.51% (p=0.008 n=5)
FastRegexMatcher/.*foo.?-12                                                                                     1913.0n ± ∞ ¹   298.8n ± ∞ ¹  -84.38% (p=0.008 n=5)
FastRegexMatcher/.?foo.+-12                                                                                      846.4n ± ∞ ¹   288.0n ± ∞ ¹  -65.97% (p=0.008 n=5)
FastRegexMatcher/foo.?|bar-12                                                                                    954.2n ± ∞ ¹   232.2n ± ∞ ¹  -75.67% (p=0.008 n=5)
geomean                                                                                                          1.258µ         421.2n        -80.97%
¹ need >= 6 samples for confidence interval at level 0.95

                                                                                                             │ before.txt  │               after.txt               │
                                                                                                             │    B/op     │    B/op      vs base                  │
FastRegexMatcher/#00-12                                                                                        0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/foo-12                                                                                        0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/^foo-12                                                                                       0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/(foo|bar)-12                                                                                  0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/foo.*-12                                                                                      0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/.*foo-12                                                                                      0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/^.*foo$-12                                                                                    0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/^.+foo$-12                                                                                    0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/.*-12                                                                                         0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/.+-12                                                                                         0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/foo.+-12                                                                                      0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/.+foo-12                                                                                      0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/foo_.+-12                                                                                     0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/foo_.*-12                                                                                     0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/.*foo.*-12                                                                                    0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/.+foo.+-12                                                                                    0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/(?s:.*)-12                                                                                    0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/(?s:.+)-12                                                                                    0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/(?s:^.*foo$)-12                                                                               0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/(?i:foo)-12                                                                                   0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/(?i:(foo|bar))-12                                                                             0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/(?i:(foo1|foo2|bar))-12                                                                       0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/^(?i:foo|oo)|(bar)$-12                                                                        0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/(?i:(foo1|foo2|aaa|bbb|ccc|ddd|e-12                                                             0.0 ± ∞ ¹   560.0 ± ∞ ¹         ? (p=0.008 n=5)
FastRegexMatcher/((.*)(bar|b|buzz)(.+)|foo)$-12                                                                0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/^$-12                                                                                         0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/(prometheus|api_prom)_api_v1_.+-12                                                            0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/10\.0\.(1|2)\.+-12                                                                            0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/10\.0\.(1|2).+-12                                                                             0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/((fo(bar))|.+foo)-12                                                                          0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/zQPbMkNO|NNSPdvMi|iWuuSoAl|qbvKM-12                                                           0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=0.444 n=5)
FastRegexMatcher/jyyfj00j0061|jyyfj00j0062|jyyfj9-12                                                           2.000 ± ∞ ¹   0.000 ± ∞ ¹  -100.00% (p=0.008 n=5)
FastRegexMatcher/(?i:(zQPbMkNO|NNSPdvMi|iWuuSoAl|-12                                                             0.0 ± ∞ ¹   560.0 ± ∞ ¹         ? (p=0.008 n=5)
FastRegexMatcher/(?i:(zQPbMkNO.*|NNSPdvMi.*|iWuuS-12                                                           4.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=0.167 n=5)
FastRegexMatcher/(?i:(.*zQPbMkNO|.*NNSPdvMi|.*iWu-12                                                           139.0 ± ∞ ¹     0.0 ± ∞ ¹  -100.00% (p=0.008 n=5)
FastRegexMatcher/fo.?-12                                                                                       0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/foo.?-12                                                                                      0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/f.?o-12                                                                                       0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/.*foo.?-12                                                                                    0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/.?foo.+-12                                                                                    0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
FastRegexMatcher/foo.?|bar-12                                                                                  0.000 ± ∞ ¹   0.000 ± ∞ ¹         ~ (p=1.000 n=5) ²
geomean                                                                                                                  ³                ?                      ³
¹ need >= 6 samples for confidence interval at level 0.95
² all samples are equal
³ summaries must be >0 to compute geomean

                                                                                                             │ before.txt  │           after.txt            │
                                                                                                             │  allocs/op  │  allocs/op   vs base           │
FastRegexMatcher/#00-12                                                                                        0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/foo-12                                                                                        0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/^foo-12                                                                                       0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/(foo|bar)-12                                                                                  0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/foo.*-12                                                                                      0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/.*foo-12                                                                                      0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/^.*foo$-12                                                                                    0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/^.+foo$-12                                                                                    0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/.*-12                                                                                         0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/.+-12                                                                                         0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/foo.+-12                                                                                      0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/.+foo-12                                                                                      0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/foo_.+-12                                                                                     0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/foo_.*-12                                                                                     0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/.*foo.*-12                                                                                    0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/.+foo.+-12                                                                                    0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/(?s:.*)-12                                                                                    0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/(?s:.+)-12                                                                                    0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/(?s:^.*foo$)-12                                                                               0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/(?i:foo)-12                                                                                   0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/(?i:(foo|bar))-12                                                                             0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/(?i:(foo1|foo2|bar))-12                                                                       0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/^(?i:foo|oo)|(bar)$-12                                                                        0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/(?i:(foo1|foo2|aaa|bbb|ccc|ddd|e-12                                                            0.00 ± ∞ ¹   18.00 ± ∞ ¹  ? (p=0.008 n=5)
FastRegexMatcher/((.*)(bar|b|buzz)(.+)|foo)$-12                                                                0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/^$-12                                                                                         0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/(prometheus|api_prom)_api_v1_.+-12                                                            0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/10\.0\.(1|2)\.+-12                                                                            0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/10\.0\.(1|2).+-12                                                                             0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/((fo(bar))|.+foo)-12                                                                          0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/zQPbMkNO|NNSPdvMi|iWuuSoAl|qbvKM-12                                                           0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/jyyfj00j0061|jyyfj00j0062|jyyfj9-12                                                           0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/(?i:(zQPbMkNO|NNSPdvMi|iWuuSoAl|-12                                                            0.00 ± ∞ ¹   18.00 ± ∞ ¹  ? (p=0.008 n=5)
FastRegexMatcher/(?i:(zQPbMkNO.*|NNSPdvMi.*|iWuuS-12                                                           0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/(?i:(.*zQPbMkNO|.*NNSPdvMi|.*iWu-12                                                           0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/fo.?-12                                                                                       0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/foo.?-12                                                                                      0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/f.?o-12                                                                                       0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/.*foo.?-12                                                                                    0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/.?foo.+-12                                                                                    0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
FastRegexMatcher/foo.?|bar-12                                                                                  0.000 ± ∞ ¹   0.000 ± ∞ ¹  ~ (p=1.000 n=5) ²
geomean                                                                                                                  ³                ?               ³
¹ need >= 6 samples for confidence interval at level 0.95
² all samples are equal
³ summaries must be >0 to compute geomean
```

### How do the optimizations work?

The core idea is that `NewFastRegexMatcher()` analyses the regexp and optimize common cases we've seen in production, across all Grafana Cloud customers. Then `MatchString()` matches the string based on applied optimizations, falling back to the golang regexp engine if no optimization was detected.

### Why new methods to `Matcher`

In this PR I also propose to add new methods to `Matcher`, specifically:

- `SetMatches()`
- `Prefix()`
- `IsRegexOptimized()`

Why?

- `IsRegexOptimized()` is just used to track the % of optimized regexp matchers (see metric above)
- `SetMatches()`: the set matches found by `FastRegexpMatcher` is an evoluted version of `findSetMatches()`. I've removed `findSetMatches()` and used `Matcher.SetMatches()` which cover more cases.
-  `Prefix()` is something we use in Mimir to further optimize the TSDB index lookup. It's unused by Prometheus right. Given it's small, I'm wondering if we can consider keeping it. Otherwise I will remove it.

### CODE REVIEW

File: tsdb/querier_test.go [hunk 1]
```diff
@@ -3304,7 +3304,7 @@ func TestPostingsForMatcher(t *testing.T) {
 		{
 			// Test case for double quoted regex matcher
 			matcher:  labels.MustNewMatcher(labels.MatchRegexp, "test", "^(?:a|b)$"),
-			hasError: true,
+			hasError: false,
```
Comment (pracucci) [inline tsdb/querier_test.go:3307]: Note to reviewers: now it's a supported case.

**Comment by yeya24** [comment]
Can we move forward the review process of this pr? It would be great to have this change merged!

**Comment by pracucci** [comment]
@bwplotka Any chance to take a look at this? The PR needs to be rebased and I will do it as soon as there's any sign of activity (to avoid having to frequently rebase a PR which doesn't move). Thanks!

**Review by bboreham — APPROVED** [review]
Apologies for taking so long to review.

The code is well-structured and fairly clear.  Also it has been running in production for many months.  I think it will be OK to accept to Prometheus.

However it is rather long, and also it depends on details of the API to Go regexps.  I think the latter cannot change under the Go 1.0 compatibility guarantees.

**Comment by yeya24** [comment]
Thanks @bboreham, can we have this pr merged?

**Comment by yeya24** [comment]
Can we resolve conflicts and merge this pr? Thanks

**Pull request merged**